### PR TITLE
Project/scalems

### PR DESCRIPTION
**Pull request by andre-merzky**
This is an aggregation of relatively small changes which get `feature/env_isolation` to work within the ScaleMS project (Frontera).  Sorry for not splitting into smaller PRs :-/

### CODE REVIEW

File: bin/radical-pilot-create-static-ve [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/sh -x
```
Comment (andre-merzky) [inline bin/radical-pilot-create-static-ve:1]: needs revert - the whole file has debug prints which need removing

File: bin/radical-pilot-create-static-ve [hunk 2]
```diff
@@ -31,7 +38,7 @@ then
     MODULES="$MODULES apache-libcloud chardet colorama idna msgpack"
     MODULES="$MODULES msgpack-python netifaces ntplib parse pymongo"
     MODULES="$MODULES pyzmq regex requests setproctitle urllib3"
-    MODULES="$MODULES python-string-utils'
+    MODULES="$MODULES python-string-utils"
```
Comment (andre-merzky) [inline bin/radical-pilot-create-static-ve:41]: This dependency is gone.

File: examples/00_getting_started.py [hunk 3]
```diff
@@ -82,8 +82,17 @@
             # create a new task description, and fill it.
             # Here we don't use dict initialization.
             td = rp.TaskDescription()
-            td.executable    = '/bin/date'
-            td.cpu_processes = 1
+            td.uid            = 'task.%03d' % i
```
Comment (andre-merzky) [inline examples/00_getting_started.py:85]: This should be reverted or cleaned.

File: examples/misc/raptor.py [hunk 4]
```diff
@@ -44,26 +40,26 @@
...
-                                        ns=session.uid)
+                                               ru.ID_CUSTOM,
+                                               ns=session.uid)
             td.executable     = "/bin/sh"
             td.cpu_threads    = cpn
             td.gpu_processes  = gpn
-            td.arguments      = [os.path.basename(master_sh), cfg_file, i]
-            td.input_staging  = [{'source': master,
-                                  'target': os.path.basename(master),
+            td.arguments      = ['./raptor_master.sh', cfg_file, i]
```
Comment (andre-merzky) [inline examples/misc/raptor.py:48]: wrapper shells should be removed.

File: examples/misc/rp_app_comm.py [hunk 5]
```diff
@@ -1,53 +0,0 @@
-#!/usr/bin/env python3
```
Comment (andre-merzky) [inline examples/misc/rp_app_comm.py:1]: ahouls bw recovered / fixed.

File: examples/misc/task_overlay.cfg [hunk 6]
```diff
@@ -0,0 +1,43 @@
+
```
Comment (andre-merzky) [inline examples/misc/task_overlay.cfg:1]: should be removed.

File: src/radical/pilot/agent/agent_0.py [hunk 7]
```diff
@@ -799,18 +792,30 @@ def _prepare_env(self, eid, env_spec):
         assert(etype == 'virtualenv')
         assert(evers)
 
-        rp_cse = 'radical-pilot-create-static-ve'
+        self._log.error('===================== find this =====================')
```
Comment (andre-merzky) [inline src/radical/pilot/agent/agent_0.py:795]: cleanup

File: src/radical/pilot/agent/agent_0.py [hunk 8]
```diff
@@ -799,18 +792,30 @@ def _prepare_env(self, eid, env_spec):
...
-        ve_cmd = '%s -p ./%s -v %s -e ". env/bs0_pre_0.sh" -m "%s"' \
-               % (rp_cse, eid, evers, ','.join(emods))
-        ru.sh_callout_bg('%s 1>%s.log 2>&1 && touch %s.ok || touch %s.fail'
-               % (ve_cmd, self.uid, eid, eid), shell=True)
+        rp_cse = ru.which('radical-pilot-create-static-ve')
+        ve_cmd = '/bin/sh -x %s -p %s/env/rp_named_env.%s -v %s ' \
+                 '-e ". env/bs0_pre_0.sh" -m "%s" | tee -a env.log 2>&1' \
+               % (rp_cse, self._pwd, env_name, evers, ','.join(emods))
+
+        self._log.debug('=== env cmd: %s', ve_cmd)
```
Comment (andre-merzky) [inline src/radical/pilot/agent/agent_0.py:802]: cleanup debug messages.

File: src/radical/pilot/agent/launch_method/base.py [hunk 9]
```diff
@@ -185,25 +186,23 @@ def get_launcher_env(self):
...
-        name = spec['name']
-        cmds = spec['cmds']
+    def get_task_named_env(self, env_name):
 
         # we assume that the launcher env is still active in the task execution
         # script.  We thus remove the launcher env from the task env before
         # applying the task env's pre_exec commands
-        tgt = '%s/env/%s.env' % (self._pwd, name)
-        self._log.debug('=== tgt : %s', tgt)
+        act = '%s/env/rp_named_env.%s.sh' % (self._pwd, env_name)
```
Comment (andre-merzky) [inline src/radical/pilot/agent/launch_method/base.py:194]: remove debug messages.

File: examples/misc/raptor.py (hunk 4 above)
Comment (andre-merzky) [reply examples/misc/raptor.py:48]: Done

File: examples/misc/task_overlay.cfg (hunk 6 above)
Comment (andre-merzky) [reply examples/misc/task_overlay.cfg:1]: done

File: src/radical/pilot/agent/launch_method/base.py (hunk 9 above)
Comment (andre-merzky) [reply src/radical/pilot/agent/launch_method/base.py:194]: Done

**Review by andre-merzky — COMMENTED** [review]
resolved  TODO items

File: src/radical/pilot/agent/resource_manager/base.py [hunk 10]
```diff
@@ -393,14 +419,109 @@ def stop(self):
...
+            self._log.debug('found %d [%d cores]', len(nodes), cores_per_node)
+            return cores_per_node.pop()
+
+        else:
+            raise ValueError('non-uniform node list, cores_per_node invalid')
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _get_node_list(self, nodes  : T_NODES,
```
[the remark this answers is not in the document]
Comment (mtitov) [reply src/radical/pilot/agent/resource_manager/base.py:508]: inside this method there is `'cores'    : [rpc.FREE] * node[1],` ([link](https://github.com/radical-cybertools/radical.pilot/blob/project/scalems/src/radical/pilot/agent/resource_manager/base.py#L518)), and `node` is taken from `nodes` (created based on nodes file), but if `rm_info.cores_per_node` was modified, then the new value is not reflected here.

I think input `nodes` should be just a list of nodes (e.g., `['node_0', 'node_1']`), but information about node parameters should be taken from rm_info
```python
                      'cores'    : [rpc.FREE] * rm_info.cores_per_node,
```

**Pull request merged**